### PR TITLE
feat(python): automagically upconvert `with_columns` kwarg expressions with multiple output names to struct; extend `**named_kwargs` support to `select`

### LINKED ISSUE #6486: Confusing (& wrong) behavior when using `with_columns` incorrectly
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Issue description

I accidentally wrote this code:

```python
import polars as pl

df = pl.DataFrame({
    'x1': [1,2,4,8,16,32],
    'x2': [1,2,3,4,5,6]
})

df.with_columns(pctChange = pl.col(['x1', 'x2']).pct_change())

>>>
shape: (6, 3)
┌─────┬─────┬───────────┐
│ x1  ┆ x2  ┆ pctChange │
│ --- ┆ --- ┆ ---       │
│ i64 ┆ i64 ┆ f64       │
╞═════╪═════╪═══════════╡
│ 1   ┆ 1   ┆ null      │
│ 2   ┆ 2   ┆ 1.0       │
│ 4   ┆ 3   ┆ 0.5       │
│ 8   ┆ 4   ┆ 0.333333  │
│ 16  ┆ 5   ┆ 0.25      │
│ 32  ┆ 6   ┆ 0.2       │
└─────┴─────┴───────────┘
```

This is the result I'd expect if I were taking the pct_change of the `x2` column, but it quietly ignores `x1`.

Two behaviours seem appropiate to me:

1. Raise an error when assigning a column using a dataframe
2. Create a struct type column.

```python
# Should behave like df.with_columns(pctChange = pl.struct(pl.col(['x1', 'x2']).pct_change()))
df.with_columns(pctChange = pl.col(['x1', 'x2']).pct_change())
>>>
shape: (6, 3)
┌─────┬─────┬────────────────┐
│ x1  ┆ x2  ┆ pctChange      │
│ --- ┆ --- ┆ ---            │
│ i64 ┆ i64 ┆ struct[2]      │
╞═════╪═════╪════════════════╡
│ 1   ┆ 1   ┆ {null,null}    │
│ 2   ┆ 2   ┆ {1.0,1.0}      │
│ 4   ┆ 3   ┆ {1.0,0.5}      │
│ 8   ┆ 4   ┆ {1.0,0.333333} │
│ 16  ┆ 5   ┆ {1.0,0.25}     │
│ 32  ┆ 6   ┆ {1.0,0.2}      │
└─────┴─────┴────────────────┘
```

In either case, the current behavior definitively violated the "don't surprise programmers" mantra. 

### Reproducible example

```python
import polars as pl

df = pl.DataFrame({
    'x1': [1,2,4,8,16,32],
    'x2': [1,2,3,4,5,6]
})

df.with_columns(pctChange = pl.col(['x1', 'x2']).pct_change())
```

### Expected behavior

Should return the same as 

```python
import polars as pl

df = pl.DataFrame({
    'x1': [1,2,4,8,16,32],
    'x2': [1,2,3,4,5,6]
})

df.with_columns(pctChange = pl.struct(pl.col(['x1', 'x2']).pct_change()))

```

Or raise an error

### Installed versions

<details>

```
---Version info---
Polars: 0.15.15
Index type: UInt32
Platform: Windows-10-10.0.22621-SP0
Python: 3.10.7 (tags/v3.10.7:6cc6b13, Sep  5 2022, 14:08:36) [MSC v.1933 64 bit (AMD64)]
---Optional dependencies---
pyarrow: 8.0.0
pandas: 1.5.2
numpy: 1.22.4
fsspec: 2022.8.2
connectorx: 0.3.1
xlsx2csv: <not installed>
matplotlib: 3.6.2
```

</details>

### PR DESCRIPTION
**Pull request by alexander-beedie**
Closes #6486, and removes the "experimental" status from `with_columns` kwargs.

**Example**:
```python
df = pl.DataFrame(
    data = {
        "x1": [1,2,6],
        "x2": [1,2,3],
    }
).with_columns(
    pct_change = pl.col(["x1","x2"]).pct_change(),
)
# ┌─────┬─────┬────────────────┐
# │ x1  ┆ x2  ┆ pct_change     │
# │ --- ┆ --- ┆ ---            │
# │ i64 ┆ i64 ┆ struct[2]      │
# ╞═════╪═════╪════════════════╡
# │ 1   ┆ 1   ┆ {null,null}    │
# │ 2   ┆ 2   ┆ {1.0,1.0}      │
# │ 6   ┆ 3   ┆ {2.0,0.5}      │
# └─────┴─────┴────────────────┘
```

### CODE REVIEW

**Comment by alexander-beedie** [comment]
@ritchie46: this is actually really nice behaviour... :)
Time to bring `with_columns` kwargs into the light for `0.16` and remove the experimental flag?

**Comment by ritchie46** [comment]
Does this work with `pl.col(DATA_TYPE)` and `pl.col(REGEX)`? How does this workout with `pl.some_expanding_expr().suffix()` and `keep_name`. 

I think we need to add a function on the rust side of py-polars that will be called if `**kwargs` are used and checks if this expression is eligible for this, I must agree, nice feature.

**Comment by alexander-beedie** [comment]
> Does this work with `pl.col(DATA_TYPE)` and `pl.col(REGEX)`? How does this workout with `pl.some_expanding_expr().suffix()` and `keep_name`.

Works like a charm for DATA_TYPE ...
```python
pl.DataFrame( {"x1":[1,2,6], "x2":[1,2,3]} ).with_columns( ints=pl.col(pl.Int64) )
# ┌─────┬─────┬───────────┐
# │ x1  ┆ x2  ┆ ints      │
# │ --- ┆ --- ┆ ---       │
# │ i64 ┆ i64 ┆ struct[2] │
# ╞═════╪═════╪═══════════╡
# │ 1   ┆ 1   ┆ {1,1}     │
# │ 2   ┆ 2   ┆ {2,2}     │
# │ 6   ┆ 3   ┆ {6,3}     │
# └─────┴─────┴───────────┘
```
...  and "expanding" calls (with suffix or keep_name) ...
```python
pl.DataFrame( {"x1":[1,2,6], "x2":[1,2,3]} ).with_columns( mins=pl.all().min().suffix("_min") )
# ┌─────┬─────┬───────────┐
# │ x1  ┆ x2  ┆ mins      │
# │ --- ┆ --- ┆ ---       │
# │ i64 ┆ i64 ┆ struct[2] │
# ╞═════╪═════╪═══════════╡
# │ 1   ┆ 1   ┆ {1,1}     │
# │ 2   ┆ 2   ┆ {1,1}     │
# │ 6   ┆ 3   ┆ {1,1}     │
# └─────┴─────┴───────────┘
df.unnest('mins')
# ┌─────┬─────┬────────┬────────┐
# │ x1  ┆ x2  ┆ x1_min ┆ x2_min │
# │ --- ┆ --- ┆ ---    ┆ ---    │
# │ i64 ┆ i64 ┆ i64    ┆ i64    │
# ╞═════╪═════╪════════╪════════╡
# │ 1   ┆ 1   ┆ 1      ┆ 1      │
# │ 2   ┆ 2   ┆ 1      ┆ 1      │
# │ 6   ┆ 3   ┆ 1      ┆ 1      │
# └─────┴─────┴────────┴────────┘

... but doesn't for REGEX (as the expr doesn't know if it will match one or more cols until evaluated).
```

**Comment by alexander-beedie** [comment]
> I think we need to add a function on the rust side of py-polars

You read my mind! I don't like relying on the `ComputeError` to detect multiple output names - being able to use something like `expr.meta.multiple_outputs()` (returning true/false) would feel much more solid.

**Comment by ritchie46** [comment]
> ... but doesn't for REGEX (as the expr doesn't know if it will match one or more cols until evaluated).

Then I think we must follow up with that checking function and then we can make it default, I think.

**Comment by alexander-beedie** [comment]
In the case of regex we can simply exec the regex on the current `df.schema`, lazy or otherwise, and find out (can't use the `expr.meta` approach as it isn't associated with a frame). I'll add this now.

**Comment by alexander-beedie** [comment]
Side-note: how does one select column names that looks like regexes? 🤣 
```python
df = pl.DataFrame( {"^x.z$":[1, 2, 3]} )
# shape: (3, 1)
# ┌───────┐
# │ ^x.z$ │
# │ ---   │
# │ i64   │
# ╞═══════╡
# │ 1     │
# │ 2     │
# │ 3     │
# └───────┘

df.select("^x.z$")
# shape: (0, 0)
# ┌┐
# ╞╡
# └┘
```

**Comment by ritchie46** [comment]
> Side-note: how does one select column names that looks like regexes? rofl
> 
> ```python
> df = pl.DataFrame( {"^x.z$":[1, 2, 3]} )
> # shape: (3, 1)
> # ┌───────┐
> # │ ^x.z$ │
> # │ ---   │
> # │ i64   │
> # ╞═══════╡
> # │ 1     │
> # │ 2     │
> # │ 3     │
> # └───────┘
> 
> df.select("^x.z$")
> # shape: (0, 0)
> # ┌┐
> # ╞╡
> # └┘
> ```

Those column names are not allowed in polars. ;) Or you must use a regex escape as fallback. 

The meta utilities are coming up btw.

**Comment by alexander-beedie** [comment]
> Those column names are not allowed in polars. ;) Or you must use a regex escape as fallback.

Shall we actively prevent them? Can easily detect/raise inside `col`.
Probably a good idea, given the above, heh.

> The meta utilities are coming up btw.

Fantastic :)

**Comment by ritchie46** [comment]
> > Those column names are not allowed in polars. ;) Or you must use a regex escape as fallback.
> 
> Shall we actively prevent them? Could easily detect/raise inside `col`. Probably a good idea, given the above, heh.
> 
> > The meta utilities are coming up btw.
> 
> Fantastic :)

Yeah, I could add an assert there.

**Comment by gab23r** [comment]
Why do not automatically convert expression with multiple names into a struct when we used alias so that :

`df = pl.DataFrame( {"x1":[1,2,6], "x2":[1,2,3]} ).with_columns(pl.col(pl.Int64).alias('ints'))` 

would work as well ?

 I think this is a nice feature, It could be extend to other functions like `select` for example

**Comment by alexander-beedie** [comment]
> Why do not automatically convert expression with multiple names into a struct when we used alias so that :

Hmm... It doesn't feel as intentional? 
(Though perhaps my feeling there will change as more people discover this method  ;)

File: py-polars/polars/internals/lazyframe/frame.py [hunk 1]
```diff
@@ -2566,7 +2566,7 @@ def with_columns(
             exprs = list(exprs)
 
         exprs.extend(
-            pli.expr_to_lit_or_expr(expr).alias(name)
+            pli.expr_to_lit_or_expr(expr, structify=True).alias(name)
```
Comment (ritchie46) [inline py-polars/polars/internals/lazyframe/frame.py:2569]: I think `structify` should only be `True` if `named_exprs` are given. The meta processing involved is relatively expensive and unneeded in many cases.

**Review by ritchie46 — COMMENTED** [review]
One small remark.

Can we also add a `structifying` example to `with_columns`. This behavior deserves some documentation. :)

File: py-polars/polars/internals/lazyframe/frame.py (hunk 1 above)
Comment (alexander-beedie) [reply py-polars/polars/internals/lazyframe/frame.py:2569]: This Is already functionally the case ;) Nothing will execute in that loop if there aren't any named_exprs, due to `for name, expr in named_exprs.items()`.

**Comment by alexander-beedie** [comment]
> Can we also add a `structifying` example to `with_columns`. This behavior deserves some documentation. :)

A very good point :) 
Will take care of that in a few hours when I'm back.

File: py-polars/polars/internals/lazyframe/frame.py (hunk 1 above)
Comment (ritchie46) [reply py-polars/polars/internals/lazyframe/frame.py:2569]: Oh, right. List comprehension. So the for is after the expression. Now I see it. :+1:

**Comment by alexander-beedie** [comment]
Done; added an extra docstring example/explanation, squashed, rebased...👌

**Comment by ritchie46** [comment]
Almost there @alexander-beedie. Maybe I missed it, but I believe we don't yet have added this functionality to `select`?

**Comment by alexander-beedie** [comment]
Ahh, you didn't miss it - apparently *I* missed that we were extending it there (though it makes perfect sense to do so!).

I'll be at the airport shortly - will see if there's time to do it before boarding - and, if not, I'm still hoping it might be possible to do a commit from 10km up 😂

**Comment by ritchie46** [comment]
> Ahh, you didn't miss it - apparently _I_ missed that we were extending it there (though it makes perfect sense to do so!).

Yeah, I want to keep those two consistent.

There is no hurry. The 0.16 release will still take few days.  Have a good flight!

**Comment by ritchie46** [comment]
Alright. Here goes! :100:

**Pull request merged**